purpose: 
train a very simple image classifier before using tensorflow which essentially answers, can a simple model learn anything from the gathered images

method:
1. load the split CSV files train, val, test, csv's
2. open each of the processed images
3. convert the images into numpy arrays
4. normalize pixel values
5. flatten images for scikit-learn
6. convert text labels into numbers
7. train logistic regression
8. evaluate on validation set

In [2]:
from pathlib import Path
from PIL import Image ## used for opening image files
import numpy as np ## used for numerical arrays since the images need to become numbers beofre the model can learn from them 
import pandas as pd ## used to load csv files 
import matplotlib.pyplot as plt

In [7]:
PROJECT_ROOT = Path("..")
PROCESSED_DIR = PROJECT_ROOT / "data" / "processed"


train_df = pd.read_csv(PROCESSED_DIR / "train.csv")
val_df = pd.read_csv(PROCESSED_DIR / "val.csv")
test_df = pd.read_csv(PROCESSED_DIR / "test.csv")

train_df.head()

,original_path,processed_path,label
0,..\data\raw\inflammatory_acne\inflam_011.jpg,..\data\processed\inflammatory_acne\inflammato...,inflammatory_acne
1,..\data\raw\comedonal_open\comeo_021.jpg,..\data\processed\comedonal_open\comedonal_ope...,comedonal_open
2,..\data\raw\inflammatory_acne\inflam_003.jpg,..\data\processed\inflammatory_acne\inflammato...,inflammatory_acne
3,..\data\raw\comedonal_open\comeo_025.jpg,..\data\processed\comedonal_open\comedonal_ope...,comedonal_open
4,..\data\raw\acne_excoriee\ae_021.jpg,..\data\processed\acne_excoriee\acne_excoriee_...,acne_excoriee


In [8]:
## takes a dataframe and returns x and y which represent image data and labels respectively
## x and y are commonly inputs/features and target labels for this project x and y are acne images and acne class names 
def load_images(df):
    ## create empty list to store all image and labels 
    images = []
    labels = []
    ## note each row contains one image path and one label 
    for _, row in df.iterrows():
        img = Image.open(row["processed_path"]).convert("RGB") ## convert rgb so that every image has 3 colour channels 
        img_array = np.array(img) / 255.0 ## convert the image into numbers for normalization
        ## note pixels are usually 0 - 255 but dividing by 255 makes it 0 to 1
        ## where 0 = black, 1 = full brightness and anything between like 0.5 is 
        ## middle brightness 
        ## need this normalization so that the model can train better since they
        ## train better when input values are small and consaistent
        ## this normalization effectively reduces the values from 0 to 255 to 0 to 1
        images.append(img_array) ## store the image array
        labels.append(row["label"]) ## store class name

    return np.array(images), np.array(labels) ## turns the python list in numpy arrays which models can work with

In [ ]:
X_train, Y_train = load_images(train_df) ## training images, training labels
X_val, Y_val = load_images(val_df) ## validation images, validation labels
X_test, Y_test = load_images(test_df) ## test images, test labels 

print("X_train shape:", X_train.shape)
print("y_train shape:", Y_train.shape)

print("X_val shape:", X_val.shape)
print("y_val shape:", Y_val.shape)

print("X_test shape:", X_test.shape)
print("y_test shape:", Y_test.shape)



X_train shape: (95, 224, 224, 3)
y_train shape: (95,)
X_val shape: (21, 224, 224, 3)
y_val shape: (21,)
X_test shape: (21, 224, 224, 3)
y_test shape: (21,)


X_train shape: (95, 224, 224, 3) represents 95 training images, 224 image height, 224 image width, 3 colour changels (rgb)
y_train shape: (95,) represents 95 labels, one for each training image

